File: Source/OldCode/Archive/2007 Phasor Protocols/PhasorProtocols/DataCellParsingState.cs
//*******************************************************************************************************
//  DataCellParsingState.cs
//  Copyright © 2009 - TVA, all rights reserved - Gbtc
//
//  Build Environment: C#, Visual Studio 2008
//  Primary Developer: James R Carroll
//      Office: PSO TRAN & REL, CHATTANOOGA - MR BK-C
//       Phone: 423/751-4165
//       Email: jrcarrol@tva.gov
//
//  Code Modification History:
//  -----------------------------------------------------------------------------------------------------
//  01/14/2005 - James R Carroll
//       Generated original version of source code.
//
//*******************************************************************************************************

namespace PCS.PhasorProtocols
{
    /// <summary>
    /// Represents the protocol independent common implementation of the parsing state used by any <see cref="IDataCell"/>.
    /// </summary>
    public class DataCellParsingState : ChannelCellParsingStateBase, IDataCellParsingState
    {
        #region [ Members ]

        // Fields
        private IConfigurationCell m_configurationCell;
        private CreateNewValueFunction<IPhasorDefinition, IPhasorValue> m_createNewPhasorValueFunction;
        private CreateNewValueFunction<IFrequencyDefinition, IFrequencyValue> m_createNewFrequencyValueFunction;
        private CreateNewValueFunction<IAnalogDefinition, IAnalogValue> m_createNewAnalogValueFunction;
        private CreateNewValueFunction<IDigitalDefinition, IDigitalValue> m_createNewDigitalValueFunction;

        #endregion

        #region [ Constructors ]

        /// <summary>
        /// Creates a new <see cref="DataCellParsingState"/> from specified parameters.
        /// </summary>
        /// <param name="configurationFrame">Reference to the <see cref="IConfigurationCell"/> associated with the <see cref="IDataCell"/> being parsed.</param>
        /// <param name="createNewPhasorValueFunction">Reference to delegate to create new <see cref="IPhasorValue"/> instances.</param>
        /// <param name="createNewFrequencyValueFunction">Reference to delegate to create new <see cref="IFrequencyValue"/> instances.</param>
        /// <param name="createNewAnalogValueFunction">Reference to delegate to create new <see cref="IAnalogValue"/> instances.</param>
        /// <param name="createNewDigitalValueFunction">Reference to delegate to create new <see cref="IDigitalValue"/> instances.</param>
        public DataCellParsingState(IConfigurationCell configurationCell, CreateNewValueFunction<IPhasorDefinition, IPhasorValue> createNewPhasorValueFunction, CreateNewValueFunction<IFrequencyDefinition, IFrequencyValue> createNewFrequencyValueFunction, CreateNewValueFunction<IAnalogDefinition, IAnalogValue> createNewAnalogValueFunction, CreateNewValueFunction<IDigitalDefinition, IDigitalValue> createNewDigitalValueFunction)
        {
            m_configurationCell = configurationCell;
            m_createNewPhasorValueFunction = createNewPhasorValueFunction;
            m_createNewFrequencyValueFunction = createNewFrequencyValueFunction;
            m_createNewAnalogValueFunction = createNewAnalogValueFunction;
            m_createNewDigitalValueFunction = createNewDigitalValueFunction;

            PhasorCount = m_configurationCell.PhasorDefinitions.Count;
            AnalogCount = m_configurationCell.AnalogDefinitions.Count;
            DigitalCount = m_configurationCell.DigitalDefinitions.Count;
        }

        #endregion

        #region [ Properties ]

        /// <summary>
        /// Gets reference to the <see cref="IConfigurationCell"/> associated with the <see cref="IDataCell"/> being parsed.
        /// </summary>
        public virtual IConfigurationCell ConfigurationCell
        {
            get
            {
                return m_configurationCell;
            }
        }

        /// <summary>
        /// Gets reference to <see cref="CreateNewValueFunction{TDefinition,TValue}"/> delegate used to create new <see cref="IPhasorValue"/> objects.
        /// </summary>
        public virtual CreateNewValueFunction<IPhasorDefinition, IPhasorValue> CreateNewPhasorValue
        {
            get
            {
                return m_createNewPhasorValueFunction;
            }
        }

        /// <summary>
        /// Gets reference to <see cref="CreateNewValueFunction{TDefinition,TValue}"/> delegate used to create new <see cref="IFrequencyValue"/> objects.
        /// </summary>
        public virtual CreateNewValueFunction<IFrequencyDefinition, IFrequencyValue> CreateNewFrequencyValue
        {
            get
            {
                return m_createNewFrequencyValueFunction;
            }
        }

        /// <summary>
        /// Gets reference to <see cref="CreateNewValueFunction{TDefinition,TValue}"/> delegate used to create new <see cref="IAnalogValue"/> objects.
        /// </summary>
        public virtual CreateNewValueFunction<IAnalogDefinition, IAnalogValue> CreateNewAnalogValue
        {
            get
            {
                return m_createNewAnalogValueFunction;
            }
        }

        /// <summary>
        /// Gets reference to <see cref="CreateNewValueFunction{TDefinition,TValue}"/> delegate used to create new <see cref="IDigitalValue"/> objects.
        /// </summary>
        public virtual CreateNewValueFunction<IDigitalDefinition, IDigitalValue> CreateNewDigitalValue
        {
            get
            {
                return m_createNewDigitalValueFunction;
            }
        }

        #endregion
    }
}

File: Source/OldCode/Archive/2006 Phasor Protocols/PhasorProtocols/DataCellParsingState.vb
'*******************************************************************************************************
'  DataCellParsingState.vb - Data frame cell parsing state class
'  Copyright © 2005 - TVA, all rights reserved - Gbtc
'
'  Build Environment: VB.NET, Visual Studio 2005
'  Primary Developer: J. Ritchie Carroll, Operations Data Architecture [TVA]
'      Office: COO - TRNS/PWR ELEC SYS O, CHATTANOOGA, TN - MR 2W-C
'       Phone: 423/751-2827
'       Email: jrcarrol@tva.gov
'
'  Code Modification History:
'  -----------------------------------------------------------------------------------------------------
'  01/14/2005 - J. Ritchie Carroll
'       Initial version of source generated
'
'*******************************************************************************************************

''' <summary>This class represents the protocol independent common implementation of the parsing state of a data frame cell that can be sent or received from a PMU.</summary>
<CLSCompliant(False)> _
Public Class DataCellParsingState

    Inherits ChannelCellParsingStateBase
    Implements IDataCellParsingState

    Private m_configurationCell As IConfigurationCell
    Private m_createNewPhasorValueFunction As IDataCellParsingState.CreateNewValueFunctionSignature(Of IPhasorDefinition, IPhasorValue)
    Private m_createNewFrequencyValueFunction As IDataCellParsingState.CreateNewValueFunctionSignature(Of IFrequencyDefinition, IFrequencyValue)
    Private m_createNewAnalogValueFunction As IDataCellParsingState.CreateNewValueFunctionSignature(Of IAnalogDefinition, IAnalogValue)
    Private m_createNewDigitalValueFunction As IDataCellParsingState.CreateNewValueFunctionSignature(Of IDigitalDefinition, IDigitalValue)

    Public Sub New( _
        ByVal configurationCell As IConfigurationCell, _
        ByVal createNewPhasorValueFunction As IDataCellParsingState.CreateNewValueFunctionSignature(Of IPhasorDefinition, IPhasorValue), _
        ByVal createNewFrequencyValueFunction As IDataCellParsingState.CreateNewValueFunctionSignature(Of IFrequencyDefinition, IFrequencyValue), _
        ByVal createNewAnalogValueFunction As IDataCellParsingState.CreateNewValueFunctionSignature(Of IAnalogDefinition, IAnalogValue), _
        ByVal createNewDigitalValueFunction As IDataCellParsingState.CreateNewValueFunctionSignature(Of IDigitalDefinition, IDigitalValue))

        m_configurationCell = configurationCell
        m_createNewPhasorValueFunction = createNewPhasorValueFunction
        m_createNewFrequencyValueFunction = createNewFrequencyValueFunction
        m_createNewAnalogValueFunction = createNewAnalogValueFunction
        m_createNewDigitalValueFunction = createNewDigitalValueFunction

        PhasorCount = m_configurationCell.PhasorDefinitions.Count
        AnalogCount = m_configurationCell.AnalogDefinitions.Count
        DigitalCount = m_configurationCell.DigitalDefinitions.Count

    End Sub

    Public Overrides ReadOnly Property DerivedType() As System.Type
        Get
            Return Me.GetType()
        End Get
    End Property

    Public Overridable ReadOnly Property ConfigurationCell() As IConfigurationCell Implements IDataCellParsingState.ConfigurationCell
        Get
            Return m_configurationCell
        End Get
    End Property

    Public Overridable ReadOnly Property CreateNewPhasorValueFunction() As IDataCellParsingState.CreateNewValueFunctionSignature(Of IPhasorDefinition, IPhasorValue) Implements IDataCellParsingState.CreateNewPhasorValueFunction
        Get
            Return m_createNewPhasorValueFunction
        End Get
    End Property

    Public Overridable ReadOnly Property CreateNewFrequencyValueFunction() As IDataCellParsingState.CreateNewValueFunctionSignature(Of IFrequencyDefinition, IFrequencyValue) Implements IDataCellParsingState.CreateNewFrequencyValueFunction
        Get
            Return m_createNewFrequencyValueFunction
        End Get
    End Property

    Public Overridable ReadOnly Property CreateNewAnalogValueFunction() As IDataCellParsingState.CreateNewValueFunctionSignature(Of IAnalogDefinition, IAnalogValue) Implements IDataCellParsingState.CreateNewAnalogValueFunction
        Get
            Return m_createNewAnalogValueFunction
        End Get
    End Property

    Public Overridable ReadOnly Property CreateNewDigitalValueFunction() As IDataCellParsingState.CreateNewValueFunctionSignature(Of IDigitalDefinition, IDigitalValue) Implements IDataCellParsingState.CreateNewDigitalValueFunction
        Get
            Return m_createNewDigitalValueFunction
        End Get
    End Property

End Class